src/main/scripts/run_e2e_test_multi_module.py

```diff
@@ -0,0 +1,134 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+import io
+import json
+import os
+import requests
+import subprocess
+import xml.etree.ElementTree as ET
+import zipfile
+import shutil
+import sys
+import argparse
+import urllib.request
+
+def run_shell_command(cmd):
+    process = subprocess.run(cmd.split(" "), stderr=subprocess.PIPE)
+    if process.returncode != 0:
+        print("Process completed with error: ", process.stderr)
+    assert process.returncode == 0
+
+# Parse command line optional arguments
+parser=argparse.ArgumentParser()
+parser.add_argument('--testRunner', help='TestRunner class to execute tests')
+parser.add_argument('--dependency', help='Path for additional dependencies to install', action='append', required=True)
+parser.add_argument('--module', help='Module for which tests need to be run', required=True)
+parser.add_argument('--framework', help='Pass this param if workflow is triggered from e2e framework repo')
+args=parser.parse_args()
+
+# Start CDAP sandbox
+print("Downloading CDAP sandbox")
+sandbox_url = "https://github.com/cdapio/cdap-build/releases/download/latest/cdap-sandbox-6.8.0-SNAPSHOT.zip"
+sandbox_dir = sandbox_url.split("/")[-1].split(".zip")[0]
+r = requests.get(sandbox_url)
+z = zipfile.ZipFile(io.BytesIO(r.content))
+z.extractall("./sandbox")
+
+print("Installing gcs connector jar")
+gcs_jar_url = "https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-hadoop2-2.2.5.jar"
+gcs_jar_fname = f"sandbox/{sandbox_dir}/lib/gcs-connector-hadoop2-2.2.5.jar"
+urllib.request.urlretrieve(gcs_jar_url, gcs_jar_fname)
+
+print("Start the sandbox")
+run_shell_command(f"chmod +x sandbox/{sandbox_dir}/bin/cdap")
+my_env = os.environ.copy()
+my_env["_JAVA_OPTIONS"] = "-Xmx24G"
+sandbox_start_cmd = "sandbox/" + sandbox_dir + "/bin/cdap sandbox restart"
+process = subprocess.Popen(sandbox_start_cmd, shell=True, env=my_env)
+process.communicate()
+assert process.returncode == 0
+
+e2e_test_module = args.module
+
+# Changedir into the plugin
+os.chdir("plugin")
+
+# Dependencies must be installed locally beforehand
+for dependency in args.dependency:
+    print(f"Installing dependency: {dependency}")
+    os.chdir(dependency)
+
+    # Get plugin artifact name and version from pom.xml.
+    root = ET.parse('pom.xml').getroot()
+    plugin_name = root.find('{http://maven.apache.org/POM/4.0.0}artifactId').text
+    plugin_version = root.find('{http://maven.apache.org/POM/4.0.0}version').text
+
+    os.chdir('target')
```

Comment:
We will need to build the dependency module first then only we can install it from the target folder right??

Reply:
yes, the entire project needs to be build before installing all dependencies and running E2E tests

Reply:
Looks like this script isn't taking care of that.

The older script first builds the plugin module and then installs it.
ref - https://github.com/cdapio/cdap-e2e-tests/blob/develop/src/main/scripts/run_e2e_test.py#L69

Reply:
Should this be the responsibility of the plugin? With multi-module projects there may be optimizations (such as only building certain modules) or using multiple threads when compiling. I wanted to leave this for the parent project to decide how to handle.

Reply:
Discussed offline to keep it as the responsibility of caller that module is built before running the script.
Can we add a comment here in the script for future reference??